### PR TITLE
test(llm): cover near-zero SMT timeout classification

### LINKED ISSUE #465: Add end-to-end test: near-zero SMT timeout yields EquivUnknown in LLM search
Deferred from #444: add a regression test that drives a near-zero `smt_timeout` through `classify()` and asserts the LLM search returns `EquivUnknown` rather than accepting the candidate or hard-failing.

Original feedback by @claude[bot] (claude-review) on PR #444 (https://github.com/pmatos/s11/pull/444#issuecomment-4736271798):

> The `verify_remaining` path in `mod.rs:192–202` is tested behaviorally by the `search_timeout_kills_slow_codex_child` test (which never reaches verification), but there's no test that passes a near-zero `smt_timeout` through `classify()` and asserts `EquivUnknown`. The `verification_config` test pins the config correctly, but a full end-to-end path from a tiny deadline to an `EquivUnknown` return would catch regressions if `EquivalenceConfig::timeout` is ever wired differently. Not blocking, but worth a future follow-up.

**Context:** `classify()` / `verify_remaining` plumbing in `src/search/llm/mod.rs:192-202` and `src/search/llm/outcome.rs`.

**Why deferred:** diminishing-returns — the reviewer explicitly flagged this as "Not blocking... worth a future follow-up", and the supplied timeout is already routed into `verification_config` (`outcome.rs`); the added coverage is a nice-to-have, not required for this PR.

_Filed automatically by `pm-autofix-pr` after dual-evaluator triage by Claude Opus 4.8 and Codex GPT-5.5._

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- exercise the LLM classifier with a valid shorter candidate and a 1 ms SMT budget
- route production `classify()` through a private injectable verifier core without changing its API or real-verifier behavior
- deterministically assert that the verifier sees the near-zero timeout and that `Unknown` maps to `EquivUnknown` with verification metrics
- remove the wall-clock-dependent carry-identity fixture and generous-budget fallback solve

## Testing

- red phase: targeted test failed to compile because `classify_with_verifier` did not exist
- green phase: targeted regression, all 7 classifier tests, and all 45 LLM tests passed
- `cargo fmt --all -- --check`
- `cargo clippy --all -- -D warnings`
- `cargo test --all`
- `./ci_check.sh`

The implementation-stage template also named `scripts/full_test.sh`, but that Vow-specific script does not exist in this s11 repository; `./ci_check.sh` is the repository-documented full gate and passed.

Closes #465

**Merge with `gh pr merge --merge` (merge commit, not squash) per CLAUDE.md.**

### CODE REVIEW

**Comment by pmatos** [comment]
Implementation note: the planned multiplication-distributivity fixture was too easy for the local Z3 build and completed within 1 ms, producing `Success` instead of exercising the timeout branch. I used eight independent carry identities (`a + b = (a ^ b) + 2*(a & b)`) instead. This preserves the intended public `classify()` seam, a positive whole-millisecond timeout, a shorter equivalent candidate, and real SMT invocation without adding a verifier mock. The final regression passed 20 consecutive local runs.

The implementation-stage template's `scripts/full_test.sh` is not present in this s11 workspace. I used the repository's documented `./ci_check.sh` full gate, plus explicit formatting, Clippy, and `cargo test --all`; all passed.

**Comment by pmatos** [comment]
## Automated code-review pass (`/code-review --fix`, xhigh)

Six findings; four applied to the working tree in 7f28583, two recorded here.

### Applied

1. **The test could not distinguish an equivalent candidate from a wrong one.** Asserting only `EquivUnknown` + `smt_called` is satisfied by a non-equivalent fixture too — substituting `sub` for `add` in the candidate left the test green. Fixed by re-classifying the same pair under a generous 30 s budget and asserting `Success`, so the 1 ms budget is provably the *only* reason for the unknown verdict. Mutation-checked: the `sub` variant now fails with `fixture must be equivalent under a generous budget, got EquivFail`.
2. **`Unknown` was not attributed to the budget.** Same fix — the generous-budget assertion rules out a non-timeout source of `SatResult::Unknown` (e.g. a future SMT-lowering incompleteness).
3. **Comment overstated the concrete fast path.** With `fast_only` off, `run_fast_path` builds its random-input mask from `live_out` only — x0–x7, which are the *outputs* of both sequences. Operands x10–x25 are never randomized and stay zero on every trial, so the fast path compares 0-vs-0 and is vacuous here. Comment rewritten to say so, and to name the generous-budget assertion as what actually pins equivalence.
4. **`smt_formula_bytes == None` contract unasserted.** Added, pinning the documented "serialize only on the unsat branch" behaviour so a refactor can't silently add a full SMT-LIB serialization to the timed-out cold path.

Cost: test runtime ~0.01 s → ~0.25 s.

### Not applied — recorded for a reviewer

5. **`timeout.as_millis() as u32` truncates sub-millisecond budgets to 0, which Z3 reads as *unbounded*** (`src/semantics/smt.rs:220`). Mechanically real: `classify(..., Duration::from_micros(999))` would run an unbounded solve and return `Success`. **But the production path is already deliberately guarded** — `SearchConfig::solver_query_timeout` (`src/search/config.rs:575-580`) returns `None` for zero and sub-ms values, with a comment anticipating this exact case ("Z3 interprets a zero timeout as unbounded... neither value may reach a backend"). So this is a defense-in-depth gap on the `pub fn classify` seam, not a live bug, and the fix is production code outside this test-only PR. Left alone rather than filed as an issue, since the repo has already reasoned about it explicitly; happy to harden `classify` with its own floor if a reviewer prefers.

6. **The test depends on Z3 wall-clock performance**, so a solver upgrade — not a code regression — could in principle turn it red. The proof costs ~250 ms against a 1 ms budget (~250x margin). The codebase's alternative pattern is deterministic injection (`TIMEOUT_PROBE_UNKNOWN`, `src/search/stochastic/mcmc.rs:517`), but there is no such seam on the `classify` path and issue #465 explicitly asked for an end-to-end test *through* `classify()`. **Margin measured rather than assumed:** the test was run 30x on a prebuilt debug binary and 30x on a prebuilt release binary — 60/60 pass, 0 flips. Note the risk is one-directional: CPU contention makes Z3 slower, which makes `EquivUnknown` more likely, not less. The measured proof cost is now recorded in the fixture comment so a future failure is diagnosable.

### Verification

`cargo fmt --check`, `cargo clippy --all-targets -D warnings`, `cargo test --lib search::llm` (45/45), and the full `./ci_check.sh` gate all pass.

**Comment by pmatos** [comment]
## `/simplify` pass — no changes applied

Ran the four-angle cleanup review (reuse, simplification, efficiency, altitude) against this branch's diff vs `origin/main` (84 added lines, all in the `mod tests` block of `src/search/llm/outcome.rs`). **Nothing was applied — the working tree is unchanged.**

### Results by angle

| Angle | Outcome |
|---|---|
| Reuse | No findings |
| Efficiency | No findings |
| Simplification | 1 finding — **skipped** (see below) |
| Altitude | No blocking findings; 1 non-blocking follow-up (see below) |

### Skipped: comment-prose trim at the second `classify()` call

The simplification pass flagged the comment above `classify(&target, &raw, &live_out, Duration::from_secs(30))` as restating the earlier vacuous-fast-path explanation. Skipped deliberately:

- The overlap is a single trailing back-reference clause ("because the fast path above never varies the operands"), not a restatement. It saves no lines to remove.
- That clause is what makes the guard **self-contained at the exact line where a future reader would delete the second `classify` call** — which is the one edit that would leave this test green while testing nothing. This is load-bearing redundancy.
- The proposed replacement text was also positionally wrong (it referred to "the 30s `classify()` call above" when the comment precedes the call).

### Non-blocking follow-up: budget-derivation chain is not pinned end to end

Recording this because it is the most likely thing to resurface in review, not as a blocker for this PR.

The issue title says *end-to-end*, but the delivered test calls `classify(..., Duration::from_millis(1))` directly with a literal near-zero budget. It does not exercise the path by which `LlmSearch::search` actually *arrives* at a near-zero SMT budget in production — `remaining_until` → `verification_timeout_for_remaining` → `config.solver_timeout_with_remaining_budget` → `classify`. That derivation chain was refactored two commits before this PR in bcbb032 (#794, "derive remaining budget from elapsed time") and remains unpinned.

Notably, "the LLM path needs the Codex CLI" is **not** the reason it stops at the unit level: `src/search/llm/mod.rs` already has a `#[cfg(all(test, unix))]` `FakeCodex` harness driving the full `search()` loop, and `SearchConfig::with_solver_timeout(Duration::from_millis(1))` pins the SMT-facing budget deterministically without wall-clock racing (as `llm_verification_timeout_is_capped_by_solver_timeout` at `mod.rs:373` already demonstrates). A loop-level version is achievable today.

It was not done here because adding a second, unix-only test in a different file is a scope expansion beyond this test-only diff, which `/simplify` explicitly excludes.

### What holds up on review

- **Placement** matches this repo's own convention — e.g. the "Issue #60 end-to-end discovery test" at `src/semantics/equivalence.rs:1935` also lives in `mod tests`. Here "end-to-end" means "real Z3, no mocks", which this test satisfies, and it runs on every platform (the `FakeCodex` alternative is `#[cfg(unix)]`-gated).
- **The fixture is not a bandaid.** `classify()` calls `check_equivalence_with_config_metrics` concretely with no trait seam to mock (unlike `mcmc.rs`'s `TimeoutProbeIsa`), so a genuinely expensive fixture is the only available route to a real `Unknown`. Adding an injection point would be a production-code change.
- **The vacuous-fast-path comment is technically accurate**, verified against source: `run_fast_path` (`equivalence.rs:576-656`) with `fast_only: false` sets `input_regs = live_out_registers` (x0-x7, the outputs only), and `generate_random_inputs`/`generate_edge_case_inputs` start from `new_zeroed()` and mutate only the passed registers — so x10-x25 are provably zero on every trial.
- **No flakiness risk.** The test was run 9x in the debug profile CI uses: 9/9 green, 0.25-0.35s per run against a 1 ms budget.

### Verification

The test's central invariant was independently confirmed before the pass: swapping `add` → `sub` in `candidate_lines` makes `near_zero_smt_timeout_classifies_as_equiv_unknown` **fail**, which is exactly what the second generous-budget `classify()` call exists to guarantee. The fixture was then restored byte-identical.

Gate on the unchanged tree: `cargo fmt -- --check` clean, `cargo test --lib search::llm::outcome` 7/7 passing.

File: src/search/llm/outcome.rs [hunk 1]
```diff
@@ -236,6 +236,69 @@ mod tests {
...
+
+        let live_out = LiveOut::from_registers(groups.map(|(output, _, _)| output).to_vec());
+        let (outcome, metrics) = classify(
+            &target,
+            &candidate_lines.join("\n"),
+            &live_out,
+            Duration::from_millis(1),
+        );
+
+        assert_eq!(outcome, IterationOutcome::EquivUnknown);
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/search/llm/outcome.rs]: Fixed in 213d4f5. Public `classify()` now delegates to a private injectable verifier core. The regression uses the same valid shorter candidate as the existing success test, asserts the verifier receives the exact 1 ms `EquivalenceConfig`, injects `EquivalenceResult::Unknown`, and verifies `EquivUnknown` plus propagated SMT metrics. This removes the real-Z3 timing dependency and the fallback 30 s solve while existing public-wrapper tests retain real-verifier coverage. Targeted classifier tests (7/7), LLM tests (45/45), `cargo clippy --all -- -D warnings`, `cargo test --all`, and `./ci_check.sh` all pass.
